### PR TITLE
Extend UnnecessarilyFullyQualified to remove `java.lang` qualifiers

### PR DESCRIPTION
**Pull request by rickie**

### CODE REVIEW

File: core/src/main/java/com/google/errorprone/bugpatterns/UnnecessarilyFullyQualified.java [hunk 1]
```diff
@@ -61,6 +61,7 @@
 public final class UnnecessarilyFullyQualified extends BugChecker
     implements CompilationUnitTreeMatcher {
 
+  private static final String JAVA_LANG = "java.lang";
```
Comment (Stephan202) [inline core/src/main/java/com/google/errorprone/bugpatterns/UnnecessarilyFullyQualified.java]: While no such package currently exist, we shouldn't also match a hypothetical `java.langSomeSuffix`:
```suggestion
  private static final String JAVA_LANG = "java.lang.";
```

File: core/src/main/java/com/google/errorprone/bugpatterns/UnnecessarilyFullyQualified.java [hunk 2]
```diff
@@ -169,12 +170,16 @@ public Void visitIdentifier(IdentifierTree identifierTree, Void aVoid) {
         continue;
       }
       List<TreePath> pathsToFix = getOnlyElement(types.values());
-      if (pathsToFix.stream()
-          .anyMatch(path -> findIdent(nameString, state.withPath(path), VAL_TYP) != null)) {
+      if (!pathsToFix.get(0).getLeaf().toString().contains(JAVA_LANG)
+          && pathsToFix.stream()
+              .anyMatch(path -> findIdent(nameString, state.withPath(path), VAL_TYP) != null)) {
         continue;
       }
```
Comment (Stephan202) [inline core/src/main/java/com/google/errorprone/bugpatterns/UnnecessarilyFullyQualified.java]: I don't think we should to special-case `java.lang` here. Instead we should generalize the `anyMatch` logic and compare any matching ident again the type symbol under consideration; if they're the same we can proceed.
```suggestion
      if (pathsToFix.stream()
          .anyMatch(
              path -> {
                Symbol ident = findIdent(nameString, state.withPath(path), VAL_TYP);
                return ident != null && !ident.equals(type);
              })) {
        continue;
      }
```

File: core/src/main/java/com/google/errorprone/bugpatterns/UnnecessarilyFullyQualified.java [hunk 3]
```diff
@@ -169,12 +170,16 @@ public Void visitIdentifier(IdentifierTree identifierTree, Void aVoid) {
...
       SuggestedFix.Builder fixBuilder = SuggestedFix.builder();
-      fixBuilder.addImport(getOnlyElement(types.keySet()).getQualifiedName().toString());
+      String newImport = getOnlyElement(types.keySet()).getQualifiedName().toString();
+      if (!newImport.contains(JAVA_LANG)) {
```
Comment (Stephan202) [inline core/src/main/java/com/google/errorprone/bugpatterns/UnnecessarilyFullyQualified.java]:
```suggestion
      if (!newImport.startsWith(JAVA_LANG)) {
```

**Review by Stephan202 — COMMENTED** [review]
Rebased and added a commit. I think we can open an upstream PR for this one. Suggested PR/commit message:
```
UnnecessarilyFullyQualified: drop `java.lang` qualifiers
```

**Review by Stephan202 — COMMENTED** [review]
Generalized the fix to cover additional cases. New suggested commit message:
```
UnnecessarilyFullyQualified: handle `java.lang` and inconsistent import usages
```

**Review by Stephan202 — APPROVED** [review]
Slightly extended suggested commit message:
```
UnnecessarilyFullyQualified: handle `java.lang` and inconsistent import usages

This introduces two improvements:
- Where possible `java.lang.SomeType` references are simplified to `SomeType`
  without introducing an associated `import` statement.
- References to `some.package.SomeType` are now also replaced with `SomeType`
  if said type is already imported and referenced by its simple name elsewhere
  in the compilation unit.
```

**Comment by rickie** [comment]
Opened PR upstream: https://github.com/google/error-prone/pull/2776

**Comment by rickie** [comment]
Closing this PR as it got merged in: https://github.com/google/error-prone/commit/c07a7a999a72552b75d05699acd39edeb812e3ce

**Pull request closed without merging**